petals-esb-deployer-impl/src/main/java/org/ow2/petals/deployer/runtimemodel/RuntimeContainer.java

```diff
@@ -140,8 +149,46 @@ public RuntimeComponent getComponent(final String id) {
         return components.get(id);
     }
 
+    /**
+     * Adding or removing from the returned collection does not affect component.
+     */
     public Collection<RuntimeComponent> getComponents() {
-        return components.values();
+        return new HashSet<>(components.values());
+    }
+
+    /**
+     * Return the RuntimeSharedLibrary with id and version from idAndVersion as <code>id:version</code>, or {code null}
+     * if not in the container.
+     * 
+     * @param idAndVersion
+     *            id:version
+     * @return the shared library if found, else null
+     */
+    public RuntimeSharedLibrary getSharedLibrary(final String idAndVersion) {
+        return sharedLibraries.get(idAndVersion);
+    }
+
+    /**
+     * 
+     * @param sharedLibrary
+     *            must not be {code null}
+     * @throws DuplicatedSharedLibraryException
+     *             shared library is already in the list and was not added
+     */
+    public void addSharedLibrary(final RuntimeSharedLibrary sharedLibrary) throws DuplicatedSharedLibraryException {
+        assert sharedLibrary != null;
+        String idAndVersion = sharedLibrary.getId() + ":" + sharedLibrary.getVersion();
+        if (sharedLibraries.containsKey(idAndVersion)) {
+            throw new DuplicatedSharedLibraryException("Shared library " + idAndVersion + " is already in the list");
+        }
+        sharedLibraries.put(idAndVersion, sharedLibrary);
+    }
+
+    /**
+     * Adding or removing from the returned collection does not affect container.
+     */
+    public Collection<RuntimeSharedLibrary> getSharedLibraries() {
+        return new HashSet<>(sharedLibraries.values());
```

Comment:
I think you can find better using class Collections.unmodifiableCollection() to avoid adding and removing from the returned collection.